java/src/test/java/com/mlt/TestUtils.java

```diff
@@ -3,123 +3,95 @@
 import static org.junit.jupiter.api.Assertions.assertEquals;
 
 import com.mlt.converter.mvt.MapboxVectorTile;
+import com.mlt.data.Feature;
 import com.mlt.data.MapLibreTile;
 import com.mlt.vector.FeatureTable;
-import java.util.Map;
+import java.util.ArrayList;
+import java.util.List;
 
 public class TestUtils {
+
+  private static int compareFeatures(
+      List<Feature> mltFeatures, List<Feature> mvtFeatures, boolean isFeatureTableSorted) {
+    int numErrors = 0;
+    for (var j = 0; j < mltFeatures.size(); j++) {
+      var mltFeature = mltFeatures.get(j);
+      var mvtFeature =
+          isFeatureTableSorted
+              ? mvtFeatures.stream().filter(f -> f.id() == mltFeature.id()).findFirst().get()
+              : mvtFeatures.get(j);
+      assertEquals(mvtFeature.id(), mltFeature.id());
+      var mltGeometry = mltFeature.geometry();
+      var mvtGeometry = mvtFeature.geometry();
+      if (!mvtGeometry.equalsExact(mltGeometry)) {
+        // System.out.println("Failure comparing geometry for feature: " + mvtFeature.id());
+        // System.out.println("    mvtGeometry: " + mvtGeometry);
+        // System.out.println("    mltGeometry: " + mltGeometry);
+        numErrors++;
+      }
+      var mltProperties = mltFeature.properties();
+      var mvtProperties = mvtFeature.properties();
+      for (var mvtProperty : mvtProperties.entrySet()) {
+        var mvtPropertyKey = mvtProperty.getKey();
+        // TODO: remove the below special case once this bug is fixed:
+        // https://github.com/maplibre/maplibre-tile-spec/issues/181
+        if (mvtPropertyKey.equals("id")) {
+          continue;
+        }
+        var mltProperty = mltProperties.get(mvtPropertyKey);
+        if (mltProperty == null) {
+          // System.out.println(mvtFeature.id() + " mlt prop missing for " + mvtProperty.getKey());
+          // System.out.println("  mvtProperties: " + mvtProperties);
+          // System.out.println("  mltProperties: " + mltProperties);
```

Comment:
Left in intentionally since these debug prints can be handy when trying to understand a failure. But commented so the current tests will not be extremely noisy (rather the number of errors is asserted on to keep things calm and readable)

Reply:
@msbarry nice suggestion. I'll consider adding that in a followup PR if this need persists. Right now I'm hoping we'll be able to solve a lot of these issues soon and revert back to asserting on the values without any debugging assistance.

Reply:
Just to double check - @msbarry are you certain `String.join(...)` would not run on every pass?  In most languages, a regular function call like `assertTrue` would get all its params computed before calling it (some languages like Rust use `macro` constructs - i.e. they do pre-compilation step to wrap the String.join in the if statement)

Reply:
Oh right I see what you're getting at. There's a version of assertEquals that takes a supplier so you can do `assertEquals(expected, actual, () -> String.join(...))` to defer generating the message if it is too expensive. https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/Assertions.html#assertEquals-byte-byte-java.util.function.Supplier-